pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/BucketDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,569 @@
+
+    private long numberDelayedMessages;
+
+    private final BucketState lastMutableBucketState;
+
+    private final TripleLongPriorityQueue sharedBucketPriorityQueue;
+
+    private final RangeMap<Long, BucketState> immutableBuckets;
+
+    private final Table<Long, Long, BucketState> snapshotSegmentLastIndexTable;
+
+    BucketDelayedDeliveryTracker(PersistentDispatcherMultipleConsumers dispatcher,
+                                 Timer timer, long tickTimeMillis,
+                                 boolean isDelayedDeliveryDeliverAtTimeStrict,
+                                 long fixedDelayDetectionLookahead,
+                                 BucketSnapshotStorage bucketSnapshotStorage,
+                                 long minIndexCountPerBucket, long timeStepPerBucketSnapshotSegment,
+                                 int maxNumBuckets) {
+        this(dispatcher, timer, tickTimeMillis, Clock.systemUTC(), isDelayedDeliveryDeliverAtTimeStrict,
+                fixedDelayDetectionLookahead,
+                bucketSnapshotStorage, minIndexCountPerBucket, timeStepPerBucketSnapshotSegment, maxNumBuckets);
+    }
+
+    BucketDelayedDeliveryTracker(PersistentDispatcherMultipleConsumers dispatcher,
+                                 Timer timer, long tickTimeMillis, Clock clock,
+                                 boolean isDelayedDeliveryDeliverAtTimeStrict,
+                                 long fixedDelayDetectionLookahead,
+                                 BucketSnapshotStorage bucketSnapshotStorage,
+                                 long minIndexCountPerBucket, long timeStepPerBucketSnapshotSegment,
+                                 int maxNumBuckets) {
+        super(dispatcher, timer, tickTimeMillis, clock, isDelayedDeliveryDeliverAtTimeStrict,
+                fixedDelayDetectionLookahead);
+        this.minIndexCountPerBucket = minIndexCountPerBucket;
+        this.timeStepPerBucketSnapshotSegment = timeStepPerBucketSnapshotSegment;
+        this.maxNumBuckets = maxNumBuckets;
+        this.cursor = dispatcher.getCursor();
+        this.sharedBucketPriorityQueue = new TripleLongPriorityQueue();
+        this.immutableBuckets = TreeRangeMap.create();
+        this.snapshotSegmentLastIndexTable = HashBasedTable.create();
+
+        this.bucketSnapshotStorage = bucketSnapshotStorage;
+
+        this.numberDelayedMessages = 0L;
+
+        this.lastMutableBucketState = new BucketState(-1L, -1L);
+    }
+
+    private void moveScheduledMessageToSharedQueue(long cutoffTime) {
+        TripleLongPriorityQueue priorityQueue = getPriorityQueue();
+        while (!priorityQueue.isEmpty()) {
+            long timestamp = priorityQueue.peekN1();
+            if (timestamp > cutoffTime) {
+                break;
+            }
+
+            long ledgerId = priorityQueue.peekN2();
+            long entryId = priorityQueue.peekN3();
+            sharedBucketPriorityQueue.add(timestamp, ledgerId, entryId);
+
+            priorityQueue.pop();
```

Comment:
I'm not sure If we exist some race condition cause poped entry is not the peeked entry. 
e.g. After the peek, another thread popped up with this entry.

Reply:
I check all use cases, they will be synchronous by `synchronized`.